### ISSUE TITLE
ACD / Hold-Out Baseline vs. Meta-Baseline

### ISSUE THREAD

**Issue by HectorAuvinen** [tqzhong/CG4MCTG#5]
Hello,

I wanted to ask something regarding the Joint-training based methods in the ACD and Hold-Out setting. Is it correct that the current training scripts always apply the meta-learning approach in the ACD and Hold-Out protocols? And reproducing the results of Table 3 and Table 4 in the paper for the non-meta-learning versions would just require removing the backup model (but this is not parametrized in the script and has to be manually modified). Thank you!

Kind regards,
Hector

**Comment by tqzhong** [comment]
Hi
You can execute this command: 
```
python dcg_meta.py --model_name_or_path 'gpt2-medium' \
 --output_dir '../ckpt' \
 --output_data_dir '../test_data' \
 --num_train_epochs 3 \
 --dataset 'Amazon' \
 --unseen_combs_path '../data/Amazon/unseen.jsonl' \
 --dataset_path '../data/Amazon/gen.jsonl' \
 --device_num 0 \
 --mode 'Hold-Out' \
 --idx 1
```
(just without enabling the ```--meta_mctg```, i.e., [1.1 Training Common MCTG](https://github.com/tqzhong/CG4MCTG/tree/main/meta-mctg) ).

**Comment by HectorAuvinen** [comment]
Hi!


Right, but when we need to apply sample training (which is the case when running e.g. Yelp in hold-out with the reported batch size), there is no check for this ```meta_mctg``` argument and no logic for separating dcg and meta_dcg. Or am I misunderstanding something here?

**Comment by tqzhong** [comment]
Hi, It might be that we didn't explain it clearly enough in the README. On top of the base arguments, add --meta_mctg and --lambda_s for meta-training, and add --meta_mctg, --lambda_s, --sample_train, and --num_sample_combs for meta-sample-training.

**Comment by HectorAuvinen** [comment]
Hello,


Looking at ```dcg_meta.py```, it seems like 

- every time you do sample training, and reach your update part (based on gradient accumulation steps) at line 559, you create the backup model and go into the meta-training loop. There is no check for the meta-argument.
- https://github.com/tqzhong/CG4MCTG/blob/f9a6f346d53bf64b66a8db7113867a0c6ed2e6ae/meta-mctg/dcg_meta.py#L559


- In the regular, non-sampling case there is a check for this argument, and based on it you either update your "normal" model or go into the meta-training loop.
https://github.com/tqzhong/CG4MCTG/blob/f9a6f346d53bf64b66a8db7113867a0c6ed2e6ae/meta-mctg/dcg_meta.py#L323


Is this correct? In this case sample training always leads to training the meta-version of each approach?

**Comment by tqzhong** [comment]
Yes, you are right. Sample-training is conducted under the premise of activating meta-mctg, and when the condition (number of training attribute combinations) is not met, it shifts from meta-training to meta-sample-training. Regular training does not use the sample-training approach because it does not involve constructing a pseudo-compositional batch.

**Comment by HectorAuvinen** [comment]
Hi!


In this case, how is it possible to produce the Hold-Out and ACD results in table 3 and 4 in the non-meta case? Running the script with the arguments

```
python dcg_meta.py --model_name_or_path 'gpt2-medium' \
 --output_dir '../ckpt' \
 --output_data_dir '../test_data' \
 --num_train_epochs 3 \
 --dataset 'Yelp' \
 --unseen_combs_path '../data/Yelp/unseen.jsonl' \
 --dataset_path '../data/Yelp/gen.jsonl' \
 --device_num 0 \
 --mode 'Hold-Out' \
 --idx 1
 ```
 
Leads to applying sample training (since the number of seen combinations is smaller than the batch size), which should not happen when using the baseline method without meta-learning. Would you be able to explain this? Thank you.

**Comment by HectorAuvinen** [comment]
Hi!


I saw that you updated ```dcg_meta.py``` to do a check for the ```meta_mctg``` argument before enabling sample training. This seems to have been behind the discrepancy I have been referring to.

The CTRL and Contrastive Prefix training scripts always apply the meta-approach. Are you planning on releasing a version where using the original baseline vs. the meta-variant is parametrized by an argument? This isn't a big change but I thought I would ask if you have a version where both methods can be used.
